### DISCUSSION TITLE
What's a csproj.in file?

### DISCUSSION THREAD

**Discussion by CeaseHarping** [google/or-tools#3022] [Q&A] [answered]
Forgive me if it's a dumb question, but what process do I follow/is required to be executed to turn the csproj.in files into csproj files so I can open the solution? It seems like they have variable placeholders inside but I've never looked so deeply at what a csproj is supposed to look like..

**Comment by Mizux** [comment, marked as answer]
When using Make of CMake based build we will generate the .Net `Foo.csproj` project file from a template call `Sample.csproj.in` (note: this is a CMake convention to add .in for "template" files).

You can find them here:
src: https://github.com/google/or-tools/tree/master/ortools/dotnet

## CMake

For variables i.e. `@FOO@` you can take a look at
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/CMakeLists.txt#L10-L13
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/cmake/dotnet.cmake#L30-L50
and we have a CMake function to generate it
`add_dotnet_example(File.cs)`
https://github.com/google/or-tools/blob/master/cmake/dotnet.cmake#L361-L422
```cmake
  configure_file(
    ${PROJECT_SOURCE_DIR}/ortools/dotnet/Sample.csproj.in
    ${DOTNET_EXAMPLE_DIR}/${EXAMPLE_NAME}.csproj
    @ONLY)
```

## Makefile
Same story but using `$(SED)` program:
variables:
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/makefiles/Makefile.dotnet.mk#L67-L74
and recipe `.csproj.in` to `.csproj`
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/makefiles/Makefile.dotnet.mk#L566-L585

BTW you should not write a csproj.in by hand and for .csproj you can copy paste one of the example archive.
todo(mizux): Document how to create a exampel project from scratch using VisualStudio "new cs project" or dotnet cli `dotnet new console...`
